### PR TITLE
[R-package] respect 'verbose' argument in lgb.cv() (fixes #4667)

### LINKED ISSUE #4667: [R-package] `lgb.cv` ignores `verbose` argument
In the R version, function `lgb.cv` has a parameter `verbose` with this description:
```
verbosity for output, if <= 0, also will disable the print of evaluation during training
```

However, regardless of what I pass for verbose, it will still print `[Info]` and higher-up messages:
```r
library(lightgbm)
data("agaricus.train")
pool = lgb.Dataset(agaricus.train$data, label=agaricus.train$label)
params = list(
    objective_type="binary",
    force_row_wise=TRUE,
    seed=123
)
lgb.cv(params, pool, verbose=0)
```
```
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.486412 -> initscore=-0.054366
[LightGBM] [Info] Start training from score -0.054366
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.476969 -> initscore=-0.092189
[LightGBM] [Info] Start training from score -0.092189
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.482957 -> initscore=-0.068198
[LightGBM] [Info] Start training from score -0.068198
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
...
```

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #4667.

On `master`, the tests added in this PR fail and the code below prints INFO and WARNING level logs.

```r
library(lightgbm)
data("agaricus.train")
dtrain <- lgb.Dataset(
    data = agaricus.train$data
    , label = agaricus.train$label
)
lgb.cv(
    params = list()
    , data = dtrain
    , verbose = -1L
    , nrounds = 3L
    , nfold = 2L
    , obj = "binary"
)
```

As of this PR, the tests pass and no logs are produced.

### CODE REVIEW

File: R-package/R/lgb.cv.R
Comment (jameslamb) [inline R-package/R/lgb.cv.R:336]: note to reviewers: I don't totally understand *why* this change fixes #4667. At first I was worried that this function might suffer from a larger problem like "`lgb.cv()` does not respect values passed to `params`", but that is not true.

Consider the following example. Setting `min_gain_to_split` to a large number for the binary objective results in no splits, as expected.

```r
library(lightgbm)
data("agaricus.train")
dtrain <- lgb.Dataset(
    data = agaricus.train$data
    , label = agaricus.train$label
)
lgb.cv(
    params = list(
        min_gain_to_split = 100000.0
    )
    , data = dtrain
    , verbose = 0L
    , nrounds = 3L
    , nfold = 2L
    , obj = "binary"
)
```

The logs are full of the following

> [LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements

Changing to `min_gain_to_split = 0.0`, those warnings go away and training produces trees with multiple leaves.

Comment (StrikerRUS) [reply R-package/R/lgb.cv.R:336]: Maybe this related to the lack of corresponding code lines in R-package?
https://github.com/microsoft/LightGBM/blob/2ef3cb81ed66706ed12282857635f0a5941d92ad/python-package/lightgbm/engine.py#L222-L225

Comment (jameslamb) [reply R-package/R/lgb.cv.R:340]: Maybe? I'm not really sure. The code in `lgb.train()` and `lgb.cv()` in the R package is very similar, and `lgb.train()` doesn't have this issue with the parameter `verbose`.

Anyway, I think this fix is still worth merging even if it isn't fully understood...I don't think adding another `$update_params()` with the passed-in params introduces any risk.

Comment (StrikerRUS) [reply R-package/R/lgb.cv.R:340]: > I don't think adding another `$update_params()` with the passed-in params introduces any risk.

Agree. But this introduces performance degradation because this method calls C API function under the hood and is being executed for each fold.

Comment (jameslamb) [reply R-package/R/lgb.cv.R:340]: > but this introduces performance degradation

Sure, but I think that in this case, the cost of calling this method once per fold (one time at set up, not on each iteration, just to be clear) is worth it in exchange for ensuring that `lgb.cv()`'s behavior is correct.

**Review by StrikerRUS — APPROVED** [review]
Despite that I'm not fully satisfied by this fix that causes performance degradation and believe there can be a way of specifying params without resetting them, I don't want to block this PR from merging. Please consider this my approval just as a comment that allows you to click "Merge" button. This isn't a "real" review because I don't understand why this fixes original issue (as you do https://github.com/microsoft/LightGBM/pull/4903#discussion_r772847332).

File: R-package/R/lgb.cv.R
Comment (StrikerRUS) [reply R-package/R/lgb.cv.R:340]: OK, I'm letting you to decide.

**Comment by jameslamb** [comment]
Ok thanks @StrikerRUS ! I'm going to try to investigate a bit more before merging this.

File: R-package/R/lgb.cv.R
Comment (jameslamb) [reply R-package/R/lgb.cv.R]: Ok looking into this more closely tonight, I think I figured out:

1. why `verbose` is not being respected
2. why it seems like ONLY `verbose` is not being respected and all other parameters are working correctly

## Short Summary

Two bugs in the R package:

1. `Dataset$slice()` calls `LGBM_DatasetGetSubset()`, which will globally set the log level to INFO if `verbosity` isn't in the parameters passed to it. And `verbosity` is never passed through.
2. When `Dataset$update_params(new_params)` is called on a `Dataset` that has already been constructed, it doesn't store `new_params` in memory in the R object (`private$params`)

I just pushed https://github.com/microsoft/LightGBM/pull/4903/commits/619eaf484232642d9e231a5186a1d2cf2c48d74a which reverts the addition of `booster$reset_parameter()` and adds fixes for these two bugs.

## Much Longer Summary

I was able to reduce the internals of `lgb.cv()` to the following code that helps narrow down the issue.

<details><summary>sample code (click me)</summary>

```r
library(lightgbm)

data("agaricus.train", package = "lightgbm")

params <- list(
    max_bin = 123
    , verbose = -7L
)

dtrain <- lgb.Dataset(
    data = agaricus.train$data
    , label = agaricus.train$label
    , params = params
)
dtest <- slice(
    dataset = dtrain
    , idxset = seq_len(100L)
)

# with a validation set, you get logs
booster_w_valid <- lightgbm:::Booster$new(
    params = params
    , train_set = dtrain
)
booster_w_valid$add_valid(data = dtest, name = "valid")
booster_w_valid$update()

# without a validation set, no logs!
booster_no_valid <- lightgbm:::Booster$new(
    params = params
    , train_set = dtrain
)
booster_no_valid$update()
```

</details>

root cause: The use of `slice()` on a `Dataset` resets LightGBM's log level globally to the default value (INFO).

More details:

1. `Dataset$slice()` gets params from `Dataset$get_params()`
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L538
2. `Dataset$get_params()` filters down to only Dataset-specific parameters.
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L583-L592
3. `verbose` is not in the Dataset-specific parameters.
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/aliases.R#L9
4. `Dataset$slice()` eventually calls `Dataset$construct()` which calls `LGBM_DatasetGetSubset()`, which runs `config.Set()`
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/src/c_api.cpp#L1407
5. `config.Set()` calls `Log::ResetLogLevel()` which resets log level **globally** for the current process.
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/src/io/config.cpp#L241-L249
6. If `verbosity` isn't found in `params`, it defaults to INFO-level
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/include/LightGBM/config.h#L543

`Dataset.subset()` in the Python package doesn't suffer from this issue because it references `self.params` (the dictionary of parameters stored as an attribute on the `Dataset` instance) instead of using `Dataset.get_params()`, so `verbose` is not filtered out.

https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/python-package/lightgbm/basic.py#L1860

At first, I tried to just fix this by adding `"verbosity"` to the list of parameters returned by `Dataset$get_params()`. Unfortunately, that can lead to a new type of problem where the value of `verbosity` stored in a constructed Dataset will take precedence over values passed to `lgb.train()` / `lgb.cv()` if that Dataset is re-used multiple times.

https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Booster.R#L42

Next, I tried just changing the use of `Dataset$get_params()` to use `Dataset$private$params` in `Dataset$slice()`. That uncovered another bug. That still suffered from the problem that once the training Dataset was constructed, verbosity was locked in to the value from the last time it was constructed.

1. When calling `Dataset$construct()`, if it's already been constructed then R calls `LGBM_DatasetUpdateParamChecking()`
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L562
2. That call is try-catched. If it fails, then `Dataset$construct()` does the right thing based on whether or not raw data is still available. However, if that call does NOT fail, `private$params` is not updated!
    * https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/R-package/R/lgb.Dataset.R#L559-L577

**Comment by jameslamb** [comment]
>  I'm going to try to investigate a bit more

Happy to say I found the root cause for this bug, and was able to revert the additional call to `Booster$reset_parameter()`! See https://github.com/microsoft/LightGBM/pull/4903#discussion_r777294852 for a detailed description.

This taught me something important about LightGBM's internals. Any time `config.Set()` is called, it will globally reset LightGBM's log level.

https://github.com/microsoft/LightGBM/blob/af5b40e1f653e25a5f3470e09de4d1c68dc59663/src/io/config.cpp#L241-L249

This becomes important considering https://github.com/microsoft/LightGBM/issues/4904#issuecomment-1003805133

> throughout LightGBM, "params" isn't used to represent "all of the configuration for LightGBM" but more like "overrides to default configuration values"

Any code creating a `config` object on the C/C++ side based on a `params_str` that doesn't include `verbosity` will end up resetting the log level to INFO for the current process, until some other calls sets it to somethingg else.

**Review by StrikerRUS — APPROVED** [review]
Brilliant investigation! I'm super excited that you've uncovered deep underlying bug. Thank you very much!

**Pull request merged**
